### PR TITLE
Fix LoftQ docs and tests

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Relates to #1525

_Don't merge this, some GPU tests are failing_

Unfortunately, the docs I wrote about how to use LoftQ were incorrect, based on a misunderstanding I had. In reality, it is quite a bit more involved to get LoftQ working, requiring a complete roundtrip first loading a non-quantized model with LoftQ, saving the LoRA weights and the modified base model, loading the just stored base model again but this time with quantization, and finally loading the LoftQ-initialized adapter on top. The docs now link to the example which demonstrates how to move through these steps.

The unit tests have been adjusted to go through these same steps but now most of them fail, i.e. the quantization error is _greater_ with LoftQ than without. This needs to be investigated.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
@yxli2123 As described above, I made a mistake in how LoftQ needs to be used. This PR adjust the docs and the unit tests to apply LoftQ correctly (from my understanding).

However, after I changed the unit tests, 8 of them started failing, only 4 passed:

```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
┃  File                        ┃  Function                                       ┃  Function Line  ┃  Error Line  ┃  Error           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━┩
│  tests/test_gpu_examples.py  │  TestLoftQ.test_bloomz_loftq_4bit[cuda]         │  1244           │  1260        │  AssertionError  │
│  tests/test_gpu_examples.py  │  TestLoftQ.test_bloomz_loftq_4bit[cpu]          │  1244           │  1260        │  AssertionError  │
│  tests/test_gpu_examples.py  │  TestLoftQ.test_bloomz_loftq_4bit_iter_5[cuda]  │  1263           │  1275        │  AssertionError  │
│  tests/test_gpu_examples.py  │  TestLoftQ.test_bloomz_loftq_4bit_iter_5[cpu]   │  1263           │  1275        │  AssertionError  │
│  tests/test_gpu_examples.py  │  TestLoftQ.test_bloomz_loftq_8bit_iter_5[cuda]  │  1293           │  1305        │  AssertionError  │
│  tests/test_gpu_examples.py  │  TestLoftQ.test_bloomz_loftq_8bit_iter_5[cpu]   │  1293           │  1305        │  AssertionError  │
│  tests/test_gpu_examples.py  │  TestLoftQ.test_t5_loftq_4bit[cuda]             │  1308           │  1320        │  AssertionError  │
│  tests/test_gpu_examples.py  │  TestLoftQ.test_t5_loftq_4bit[cpu]              │  1308           │  1320        │  AssertionError  │
└──────────────────────────────┴─────────────────────────────────────────────────┴─────────────────┴──────────────┴──────────────────┘
```

In these unit tests, I measure the model output from the base model without any quantization, then the model output of the bnb-quantized model, and finally the model output of the quantized model using LoftQ. The expectation is that the model output with LoftQ should be closer to the base model output than the bnb-quantized model output (without LoftQ). 

(Note that I'm only testing the LoftQ should lead to closer results, even if it's only marginally closer. Ideally, however, we want it be X% better, i.e. there should be a certain margin)

I'm not sure why the error with LoftQ is greater and not smaller. Am I still applying LoftQ incorrectly? Please check the unit test if I make some mistake there. For debugging purposes, I added a check that on each individual layer, the [residual error](https://github.com/huggingface/peft/blob/34f3fba2b3f76d62a55491524301baca83dbc22b/src/peft/utils/loftq_utils.py#L220) does indeed decrease when applying the LoRA weights initialized with LoftQ, so this part seems to be correct.

Moreover, while working on this, some questions came up:

1. Why do we need to override the weights with the dequantized weights ([here](https://github.com/huggingface/peft/blob/34f3fba2b3f76d62a55491524301baca83dbc22b/src/peft/tuners/lora/layer.py#L170)). Why can we not just load the quantized base model with the LoftQ-LoRA weights, i.e. skip [this step](https://github.com/huggingface/peft/blob/34f3fba2b3f76d62a55491524301baca83dbc22b/examples/loftq_finetuning/quantize_save_load.py#L176-L177)? I tried that as well but test are failing no matter what.
2. Is it really working correctly when we pass `num_iter > 1`? I feel like [this line](https://github.com/huggingface/peft/blob/34f3fba2b3f76d62a55491524301baca83dbc22b/src/peft/utils/loftq_utils.py#L225) is not correct, because we are iteratively changing the residual but `lora_A` and `lora_B` are only determined by the very last step. Maybe those need to be incremented iteratively as well?

**Review by pacman100 — COMMENTED** [review]
Thank you Benjamin. After going through the tests, the examples for LoftQ. I believe that the discrepancy and failing tests happen due to the casting of activation to `float16` by bitsandbytes when loading the loftq base and adapters.

**Comment by pacman100** [comment]
So, the difference between original weight W and quantized+adapters weights (Q+BA) would be less when compared to just using Q, it isn't reflecting in logits because activations at each layer are being casted to `float16` by bnb.

**Review by pacman100 — APPROVED** [review]
Thank you @BenjaminBossan for all the work on rectifying LoftQ tests and uncovering best practices for applying it!
